app/database/build_store.py

```diff
@@ -0,0 +1,537 @@
+        builds = []
+        for build_id in range(start + 1, end + 1):
+            try:
+                builds.append(self.get(build_id, allow_incompleted_builds=allow_incompleted_builds))
+            except IncompleteBuild:
+                builds.append(None)
+        return builds
+
+    def add(self, build: Build):
+        """
+        Add new build to collection.
+        :param build: The build to add to the store.
+        """
+        build_id = self._store_build(build)
+        build._build_id = build_id
+        self._cached_builds_by_id[build_id] = build
+
+    def save(self, build: Build):
+        """
+        Save current state of given build.
+        We assume that this build already exists in the database. You should always call `add` before
+        trying to ever call a `save`. This should already be done for you in cluster_master.
+        We also assume this build is already in the cache in its current state.
+        :param build: The build to save to database.
+        """
+        self._logger.notice('Saving build (id: {}) in database.'.format(build.build_id()))
+        self._update_build(build)
+
+    def clean_up(self):
+        """
+        Save current state of all cached builds.
+        """
+        for build_id in self._cached_builds_by_id:
+            build = self._cached_builds_by_id[build_id]
+            self._update_build(build)
+        self._session.commit()
+
+    def count_all_builds(self) -> int:
+        """
+        Return the total amount of builds stored in the database.
+        """
+        return self._session.query(func.count('*')).select_from(BuildStateSchema).scalar()
+
+    def count_cached_builds(self) -> int:
+        """
+        Return the amount of builds stored in the cache.
+        """
+        return len(self._cached_builds_by_id)
+
+    def _store_build(self, build: Build) -> int:
+        """
+        Serialize a Build object and commit all of the parts to the database, and then
+        return the build_id that was assigned after committing.
+        :param build: The build to store into the database.
+        """
+        # Build Status
+        build_schema = BuildStateSchema(
+            completed=build._status() == BuildState.FINISHED
+        )
+        self._session.add(build_schema)
+
+        # Commit this first to get the build_id created by the database
+        # We use this build_id to store the other parts of a Build object
+        self._session.commit()
+        build_id = build_schema.build_id
+
+        # Build Meta
+        build_meta = BuildMetaSchema(
+            build_id=build_id,
+            artifacts_tar_file=build._artifacts_tar_file,
+            artifacts_zip_file=build._artifacts_zip_file,
+            error_message=build._error_message,
+            postbuild_tasks_are_finished=bool(build._postbuild_tasks_are_finished),
+            timing_file_path=build._timing_file_path,
```

Comment:
Some of these things (especially the paths) seem like they could be computed from the build id and don't need to be saved.